assets/js/Model.js

```diff
@@ -7,23 +7,87 @@ global.OO = require( '../../node_modules/oojs/dist/oojs.js' );
  */
 App.Model = function appModel( translations ) {
 	OO.EventEmitter.call( this );
-	this.translations = translations;
+	this.localStorageKey = 'svgtranslate';
+	this.translations = translations || {};
 	this.sourceLang = null;
+	this.targetLang = null;
+	this.langs = this.getLangs( this.translations );
 };
 
 OO.mixinClass( App.Model, OO.EventEmitter );
 
+/**
+ * Get all available languages (note that translations may not be complete for all).
+ *
+ * @param {Object} translations
+ * @return {Array}
+ */
+App.Model.prototype.getLangs = function ( translations ) {
+	var t, l, langs, lang,
+		tspans = Object.keys( translations ),
+		out = [];
+	for ( t = 0; t < tspans.length; t++ ) {
+		langs = Object.keys( translations[ tspans[ t ] ] );
+		for ( l = 0; l < langs.length; l++ ) {
+			lang = langs[ l ];
+			// If it's not already in the list, add it.
+			if ( out.indexOf( lang ) === -1 ) {
+				out.push( lang );
+			}
+		}
+	}
+	return out;
+};
+
+/**
+ * Set the source language.
+ *
+ * @param {string} lang The source language code.
+ */
 App.Model.prototype.setSourceLang = function ( lang ) {
+	// Make sure the source lang exists.
+	if ( this.langs.indexOf( lang ) === -1 ) {
+		return;
+	}
+	// If it's difference from the current value, change it.
 	if ( lang !== this.sourceLang ) {
 		this.sourceLang = lang;
 		this.emit( 'sourceLangSet' );
+		this.saveToLocalStorage( 'sourceLang', lang );
```

Comment:
no biggie, but just logic-wise, it might be better to emit the event *after* the saving happens. This would ensure that anything else listening to this event can trust that the data in local storage and model are the same.

That said, nothing outside the model should need to read local storage directly anyways... the model should be the only thing that directly does it, and everything else should trust the model alone, but still, I think it makes more logical sense here to do things the other order.

Reply:
Good point. Done. And the two localstorage methods are marked as protected (not that that's a hard enforcement, but it's a hint to future devs).